scio-core/src/main/scala/com/spotify/scio/coders/instances/AvroCoders.scala

```diff
@@ -128,11 +128,26 @@ trait AvroCoders {
 
   implicit def avroSpecificRecordCoder[T <: SpecificRecord: ClassTag]: Coder[T] = {
     val clazz = ScioUtil.classOf[T]
+
+    def getSchemaFromEmptyInstance = try {
+      clazz.getDeclaredConstructor().newInstance().getSchema
+    } catch {
+      // when T is abstract class or interface it is not possible to retrieve a schema
+      case cause: Throwable =>
+        val msg =
+          "Failed to create a coder for SpecificRecord because it is impossible to retrieve an " +
+            s"Avro schema by instantiating $clazz. Use only a concrete type implementing " +
+            s"SpecificRecord or use GenericRecord type in your transformations if a concrete " +
+            s"type is not known in compile time."
+        throw new Throwable(msg, cause)
+    }
+
     // Try to get the schema with SpecificData.getSchema
     // This relies on private SCHEMA$ field that may not be defined on custom SpecificRecord instance
-    // Otherwise create a default instance and call getSchema
     val schema = Try(SpecificData.get().getSchema(clazz))
-      .getOrElse(clazz.getDeclaredConstructor().newInstance().getSchema)
+      // Otherwise create a default instance and call getSchema
+      .getOrElse(getSchemaFromEmptyInstance)
```

Comment:
`orElse` will make this more idiomatic IMHO
```suggestion
      .orElse(Try(clazz.getDeclaredConstructor().newInstance().getSchema))
      .getOrElse {
        val msg = ...
        throw new RuntimeException(msg)
      }
```